indigo_api/models/works.py

```diff
@@ -269,6 +269,26 @@ def all_commenceable_provisions(self, date=None):
 
         return provisions
 
+    def uncommenced_provision_ids(self, date=None):
```

Comment:
I'm trying to work out why we can't use something like `[p.id for p in self.uncommenced_provisions(date)]`. It seems odd that we have `uncommenced_provisions()` and `commenced_provisions()` and yet the id-based versions don't make use of them, and that `has_uncommenced_provisions()` isn't just `bool(self.uncommenced_provisions())` or similar.

Basically, we have four functions that do very similar work in very slightly different ways.

Reply:
Agreed. I'm thinking to nuke `uncommenced_provisions()` entirely, because all we ever need is a list of the ids.
The reason we can't use `[p.id for p in self.uncommenced_provisions(date)]` is `uncommenced_provisions()` would return a tree structure if we were to update it to work correctly with `all_commenceable_provisions()`, so we'd have to recurse into that.
For `has_uncommenced_provisions()`, I thought it might simplify things by stopping at the first uncommenced provision rather than going through the whole commencement tree multiple times. I'll have another look, maybe we should just `use all_uncommenced_provision_ids()` as a proxy instead.

Reply:
Cool. I suspect with a helper like `descend_toc_pre_order` a lot of this becomes much easier, and so having dedicated functions to do things like extract ids, isn't required.